### PR TITLE
perf(NODE-4727): Improve performance of buffering and cursors

### PR DESCRIPTION
**Pull request by nbbeeken**
#### What is changing?

Anywhere we used arrays or an external dependency I've migrated us to our own linked list implementation.

#### What is the motivation for this change?

In v4 the cursor implementation was changed to shift documents from the front of a list which has a negative impact of moving every item in the list up by one index. I considered tracking an offset on the cursor, but after wrestling with a number of tests that reveled the many locations the index would need to be incremented and decremented to correctly track the next document, it seemed worth using a data structure fitting the job.

##### LinkedList

I offer our own implementation of linked list which is relatively simple to follow and for us to make future adjustments to if needed. It is circularly linked with an always defined sentinel head node to remove the need for null checks on next/prev. I've added unit test benchmarks, I'll post in a follow up comment.

##### BufferPool

The linked list now backs our BufferPool implementation so that buffers can be removed from the head of list as needed for reading. I removed the logic for consuming / not consuming buffers as it added unneeded complexity to the read implementation. The non consume read was being used to read an int32 of the top of the current buffers, instead I've added a function specialized for that job, and read will always consume from the pool. 

##### AbstractCursor

We now pay an upfront cost of constructing a list from the array returned to us by BSON (ideally BSON could be asked to build this List), one iteration through the documents to make a List so we can remove documents from the front. This improves the time it takes to `shift` and keeps the memory usage of a cursor consistent with what we had prior to this change. (Keeping an offset meant also maintaining the entire array of documents in the cursor until the nextBatch was received) 

##### SessionPool & ConnectionPool & ServerSelection

The above all do not require random access to the data they manage, migrating to a LL seems like an easy win, but I can easily roll these changes back. 

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

**Review by dariakp — CHANGES_REQUESTED** [review]

File: src/utils.ts [hunk 1]
```diff
@@ -898,97 +898,224 @@ export function deepCopy<T>(value: T): T {
 }
 
 /** @internal */
-const kBuffers = Symbol('buffers');
-/** @internal */
-const kLength = Symbol('length');
+type ListNode<T> = { value: T; next: ListNode<T>; prev: ListNode<T> };
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/utils.ts:901]: I think Neal's approach is correct.  With the exception of during initialization, a Node will never have `null` `next` or `prev`s.  If we adjust the `ListNode` type to account for null pointers, then we need to check for null everywhere we use them.

File: src/utils.ts [hunk 2]
```diff
@@ -898,97 +898,224 @@ export function deepCopy<T>(value: T): T {
...
+      prev: this.head as unknown as ListNode<T>,
+      value
+    };
+    this.head.next.prev = newNode;
+    this.head.next = newNode;
+  }
+
+  private remove(node?: ListNode<T> | null) {
+    if (node == null || this.length === 0) {
+      return { next: null, prev: null, value: null };
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/utils.ts:989]: I think we would just return `null` here, no?  The return type is `ListNode<T> | null`.

File: test/benchmarks/unitBench/list.bench.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
+/* eslint-disable no-restricted-modules */
+const chalk = require('chalk');
+const { List } = require('../../../lib/utils');
+const { createHistogram } = require('perf_hooks');
+
+const iterations = 100;
+const defaultItemsSize = 100000;
+const makeBigArray = (length = defaultItemsSize) => Array.from({ length }).fill(1);
```
Comment (baileympearson) [inline test/benchmarks/unitBench/list.bench.js:8]: You use arrow functions here and functions below.  Can we choose one and be consistent throughout the file?

My preference is always `function`s but it's up to you.

File: test/benchmarks/unitBench/list.bench.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  process.stdout.write(
+    `  ${chalk.cyan('p95th')}:  ${chalk.red(makeReadableTime(h.percentile(95)))} ms |`
+  );
+  process.stdout.write(
+    `  ${chalk.cyan('p99th')}:  ${chalk.red(makeReadableTime(h.percentile(99)))} ms |`
+  );
+  console.log('\n' + '-'.repeat(155));
+};
+
+void (function testArrayShift() {
```
Comment (baileympearson) [inline test/benchmarks/unitBench/list.bench.js:30]: Any reason for the IIFEs?  Why not just throw them inside a `main` and call the `main` function?

File: src/utils.ts [hunk 5]
```diff
@@ -898,97 +898,224 @@ export function deepCopy<T>(value: T): T {
...
+    const newNode: ListNode<T> = {
+      next: this.head.next,
+      prev: this.head as unknown as ListNode<T>,
+      value
+    };
+    this.head.next.prev = newNode;
+    this.head.next = newNode;
+  }
+
+  private remove(node?: ListNode<T> | null) {
```
Comment (baileympearson) [inline src/utils.ts:987]: Let's group all public methods together and all private methods together.  idc where they each go but in the past, I've seen the following structure for classes (top to bottom in a file):

- private variables
- public variables
- getters/setters
- constructor
- private methods
- public methods

whatever order is fine, but let's organize it well 🙂

File: src/utils.ts [hunk 6]
```diff
@@ -898,97 +898,224 @@ export function deepCopy<T>(value: T): T {
...
+  }
+
+  first(): T | null {
+    // If the list is empty, value will be the sentinel's null
+    return this.head.next.value;
+  }
+
+  last(): T | null {
+    // If the list is empty, value will be the sentinel's null
+    return this.head.prev.value;
```
Comment (baileympearson) [inline src/utils.ts:1037]: personal preference: instead of relying on the comment to explain the clever code, we could make the code self-documenting and the comment unnecessary if we wrote it like this:

```typescript
return this.length === 0 ? null : this.head.prev.value;
```

same comment applies to `first()`.

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: src/utils.ts (hunk 6 above)
Comment (nbbeeken) [reply src/utils.ts:1036]: I don't think adding code helps document something, I can further explain what I mean in the comment but adding code is adding steps for no functional gain, we have a null saved at `head.value` so we might as well use it

File: src/utils.ts [hunk 7]
```diff
@@ -898,97 +898,224 @@ export function deepCopy<T>(value: T): T {
...
 
 /**
  * A pool of Buffers which allow you to read them as if they were one
  * @internal
  */
 export class BufferPool {
-  [kBuffers]: Buffer[];
-  [kLength]: number;
+  private buffers: List<Buffer>;
+  private totalByteLength: number;
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/utils.ts:1046]: thanks! just for anyone else wondering about this, since this is internal to our driver there isn't a reason to keep properties any more private than TS enforced private, and symbols add a slim amount of overhead that isn't necessary for protecting this class from outside edits at runtime.

File: src/utils.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/utils.ts]: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/first.html
https://ruby-doc.org/core-2.4.1/Array.html#method-i-first
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.first

I've used plenty of examples of first/last in my past so I'll leave it as it, I've added a tsdoc to help clarify the methods.

File: src/utils.ts [hunk 8]
```diff
@@ -898,97 +898,224 @@ export function deepCopy<T>(value: T): T {
...
   }
 
-  /** Returns the requested number of bytes without consuming them */
-  peek(size: number): Buffer {
-    return this.read(size, false);
+  /**
+   * If BufferPool contains 4 bytes or more construct an int32 from the leading bytes,
+   * otherwise return null. Size can be negative, caller should error check.
+   */
+  readSize(): number | null {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/utils.ts:1067]: It's specific to reading an int32 from the front of a buffer not OpMsg (the same logic goes for OpQuery). Grabbing a leading int32 is a common enough operation in our line of work that I don't think it adds bloat to keep it in this class, we only ever call read with a value based on the int32 that is at the leading 4 bytes of the current buffers

File: src/cursor/abstract_cursor.ts [hunk 9]
```diff
@@ -282,7 +283,17 @@ export abstract class AbstractCursor<
 
   /** Returns current buffered documents */
   readBufferedDocuments(number?: number): TSchema[] {
-    return this[kDocuments].splice(0, number ?? this[kDocuments].length);
+    const bufferedDocs: TSchema[] = [];
+    const documentsToRead = number ?? this[kDocuments].length;
+
+    for (let count = 0; count < documentsToRead; count++) {
+      const document = this[kDocuments].shift();
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:290]: > we `splice` ed the array, which did not modify

That's not correct, [splice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) modifies the array you use it on, so we were removing elements from the `kDocuments` array and we still are now.

> Why are we filtering out null values now? is that to handle the sentinel?

No the sentinel isn't exposed outside the list, it's because of typescript, even though `documentsToRead` is bound to the size of the list and you know, and I know, and everyone who can read the code knows it cannot be null... typescript cannot know :sadcat: `shift` returns `T | null` so I must either cast away the problem or conditionally check its existence.

File: src/cursor/abstract_cursor.ts [hunk 10]
```diff
@@ -687,22 +698,14 @@ export abstract class AbstractCursor<
...
+function nextDocument<T>(cursor: AbstractCursor<T>): T | null {
   const doc = cursor[kDocuments].shift();
-  if (doc) {
-    const transform = cursor[kTransform];
-    if (transform) {
-      return transform(doc) as T;
-    }
 
-    return doc;
+  if (doc != null && cursor[kTransform]) {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts]: Ah mistake, TS encouraged me to check nullishness I'm in the habbit of being explicit, I'll revert

File: src/cursor/abstract_cursor.ts [hunk 11]
```diff
@@ -733,7 +736,7 @@ export function next<T>(
     return callback(undefined, null);
   }
 
-  if (cursor[kDocuments] && cursor[kDocuments].length) {
+  if (cursorId != null && cursor[kDocuments].length !== 0) {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts]: Not sure... removed.

File: test/benchmarks/unitBench/list.bench.js (hunk 3 above)
Comment (nbbeeken) [reply test/benchmarks/unitBench/list.bench.js:8]: sure, made them consistent

File: test/benchmarks/unitBench/list.bench.js (hunk 4 above)
Comment (nbbeeken) [reply test/benchmarks/unitBench/list.bench.js]: Added a main

File: test/benchmarks/unitBench/list.bench.js [hunk 12]
```diff
@@ -0,0 +1,109 @@
+/* eslint-disable no-restricted-modules */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/benchmarks/unitBench/list.bench.js:1]: Perhaps, but to some extent we are testing nodejs / v8 and another library, I'm in favor of testing lets file something that captures the array of variables there.

File: test/unit/utils.test.ts [hunk 13]
```diff
@@ -56,44 +57,105 @@ describe('driver utils', function () {
...
-      expect(buffer).property('length').to.equal(4);
+    it('should have a readonly length', () => {
+      // @ts-expect-error: checking for readonly runtime behavior
+      expect(() => (new BufferPool().length = 3)).to.throw(TypeError);
     });
 
-    context('peek', function () {
-      it('exact size', function () {
+    describe('readSize', function () {
+      it('less than an int32 sized buffer returns null', () => {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/unit/utils.test.ts]: updated titles!

File: test/unit/utils.test.ts [hunk 14]
```diff
@@ -128,9 +190,233 @@ describe('driver utils', function () {
...
+
+      beforeEach(() => {
+        list = new List();
+        // Just to make pushing not part of the tests here
+        list.push(1);
+        list.push(2);
+        expect(Array.from(list)).to.deep.equal([1, 2]);
+      });
+
+      it('should take an item to the end of a list', () => {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/unit/utils.test.ts:376]: done!

File: test/unit/utils.test.ts [hunk 15]
```diff
@@ -128,9 +190,233 @@ describe('driver utils', function () {
...
+
+      beforeEach(() => {
+        list = new List();
+        // Just to make pushing not part of the tests here
+        list.push(1);
+        list.push(2);
+        expect(Array.from(list)).to.deep.equal([1, 2]);
+      });
+
+      it('should take an item to the start of a list', () => {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/unit/utils.test.ts]: done

File: test/unit/utils.test.ts [hunk 16]
```diff
@@ -128,9 +190,233 @@ describe('driver utils', function () {
...
+
+  describe('class List', () => {
+    describe('constructor()', () => {
+      it('should make an empty list', () => {
+        const list = new List();
+        expect(list).to.have.property('length', 0);
+        // Double checking some internals, if future code changes modify these expectations
+        // They are not intended to be set in stone or expected by users of the List class
+        expect(list).to.have.property('head').that.is.not.null;
+        expect(list).to.have.nested.property('head.next').that.is.not.null;
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/unit/utils.test.ts]: Indeed, and I aligned this with the assertions in `clear()`

File: test/unit/utils.test.ts [hunk 17]
```diff
@@ -128,9 +190,233 @@ describe('driver utils', function () {
...
+    describe('get length', () => {
+      it('should be readonly', () => {
+        const list = new List<number>();
+        expect(() => {
+          // @ts-expect-error: testing readonly-ness
+          list.length = 34;
+        }).to.throw(TypeError);
+      });
+
+      it('should return the count of items in the list', () => {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/unit/utils.test.ts]: broken them up and added a test for not falling below zero

File: src/utils.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/utils.ts:1067]: Okay, if we keep it let's name it `readInt32`, not `readSize`

File: src/utils.ts (hunk 5 above)
Comment (baileympearson) [reply src/utils.ts]: My preference is by public / private because it's easier to understand the public API of the class

File: src/sessions.ts [hunk 18]
```diff
@@ -939,14 +940,7 @@ export class ServerSessionPool {
...
 
-    while (this.sessions.length) {
-      const pooledSession = this.sessions[this.sessions.length - 1];
-      if (pooledSession.hasTimedOut(sessionTimeoutMinutes)) {
-        this.sessions.pop();
-      } else {
-        break;
-      }
-    }
+    this.sessions.prune(session => session.hasTimedOut(sessionTimeoutMinutes));
```
[the remark this answers is not in the document]
Comment (dariakp) [reply src/sessions.ts:943]: it would be nice to have a proper test for this, if it's not too much trouble

File: src/utils.ts [hunk 19]
```diff
@@ -898,97 +898,223 @@ export function deepCopy<T>(value: T): T {
...
+    const newNode: ListNode<T> = {
+      next: this.head.next,
+      prev: this.head,
+      value
+    };
+    this.head.next.prev = newNode;
+    this.head.next = newNode;
+  }
+
+  private remove(node?: ListNode<T> | null): T | null {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/utils.ts:987]: I don't think its necessary for the cases we use the list for, nullish values work inside this list, just not null and that is intentional

File: src/utils.ts [hunk 20]
```diff
@@ -898,97 +898,224 @@ export function deepCopy<T>(value: T): T {
...
       return Buffer.alloc(0);
     }
 
-    let result: Buffer;
-
-    // read the whole buffer
-    if (size === this.length) {
-      result = Buffer.concat(this[kBuffers]);
+    // We know we have enough, we just don't know how it is spread across chunks
+    // TODO(NODE-XXXX): alloc API should change based on raw option
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/utils.ts]: added!

File: src/utils.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/utils.ts]: done

File: src/cursor/abstract_cursor.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:290]: We did not have adequate tests for this API, I've added more and fixed the bounding

File: src/utils.ts (hunk 6 above)
Comment (nbbeeken) [reply src/utils.ts:1036]: This is a matter of taste, I don't think it is particularly clever to use a variable we have access to and we test for. This is currently branchless logic, while the developer has to read a bit to catch the case they may not have considered (and the comment intends to help them with that) it is less runtime steps to accomplish the goal of the method

File: src/utils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dariakp) [reply src/utils.ts:901]: This would be a fine argument if the ListNode<T> was returned to the users, but it's actually just used for the internal implementation of the list, and saying that the things that are currently typed as `ListNode<T>` always have something in them when the list methods interact with them is simply not true. The point of the TS type here is to help make sure that the internal implementation of the list accounts for all possible values of ListNode<T>. My suggestion is to explicitly define `head: HeadNode<T> | NullNode` where `HeadNode<T> = { value: null; next: ListNode<T>; prev: ListNode<T> }` and `NullNode = { value: null; next: NullNode; prev: NullNode }`. It would make it more clear that the head never contains an actual value of its own. Furthermore, setting `prev:` and `next:` of `ListNode<T>` to be `ListNode<T> | HeadNode<T>` would likewise account for those scenarios of the node in question being at the start or end of the list.

File: test/unit/utils.test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (dariakp) [reply test/unit/utils.test.ts:376]: you only did it on shift, not pop

File: src/utils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/utils.ts:901]: I've added some better typing, so now that our head node is special it moves some of the casting but I think it improves the understanding here that head is special and holds null and the rest are the items in the list. 

> `NullNode = { value: null; next: NullNode; prev: NullNode }`

Trying this doesn't get type value. In theory `node !== this head` should narrow away the HeadNode to make node only a ListNode, thus making it's value the correct T, but this reference check isn't something the language supports. (Unless I'm missing something)

File: test/unit/utils.test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/unit/utils.test.ts]: Good catch, sorry! fixed

File: src/utils.ts (hunk 6 above)
Comment (nbbeeken) [reply src/utils.ts:1040]: With the TS suggestions Daria has made the type of head is now correctly null so that lends more documentation to this behavior

File: src/utils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/utils.ts]: From slack: some more improvements were just pushed to rid us of some casting and the casting that is left is clear where the edge case is with narrowing the type

File: test/unit/sessions.test.js [hunk 21]
```diff
@@ -428,20 +428,28 @@ describe('Sessions - unit', function () {
       done();
     });
 
-    it('should remove sessions which have timed out on release', function (done) {
+    it('should remove all sessions which have timed out on release', function () {
```
Comment (dariakp) [inline test/unit/sessions.test.js:431]: let's make sure that we test that old sessions are removed regardless of where they are in the list of sessions, the following four combinations test distinct looping behaviors that I think should cover it:
1) old, new [old sessions at the start]
2) new, old [old sessions at the end]
3) new, old, new [old sessions in the middle]
4) old, new, old [non-contiguous old sessions]

**Review by dariakp — CHANGES_REQUESTED** [review]

**Review by dariakp — CHANGES_REQUESTED** [review]

File: test/unit/sessions.test.js [hunk 22]
```diff
@@ -428,20 +428,105 @@ describe('Sessions - unit', function () {
...
+        expect(pool.sessions).to.have.lengthOf(3);
+        const anyTimedOutSessions = pool.sessions
+          .toArray()
+          .some(s => s.hasTimedOut(30 * 60 * 1000));
+        expect(anyTimedOutSessions, 'Unexpected timed out sessions found in pool after release').to
+          .be.false;
+      });
+    });
+
+    it('should remove all sessions which have timed out on release', function () {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/unit/sessions.test.js]: removed

**Review by dariakp — APPROVED** [review]
